blogposts/2021/optimizing-a-code-intelligence-commit-graph.md

```diff
@@ -0,0 +1,699 @@
+
+We want to search the graph in both directions, so we perform the operation again but visit the commits in the reverse order. The set of forward-visible uploads and backwards-visible uploads can then be merged using the same rules as stated above. The complete set of visible uploads for each commit for this example commit graph are shown below.
+
+| Commit              | Descendant visibility         | Ancestor visibility           | Combined visibility           | Nearest upload |
+| ------------------- | ----------------------------- | ----------------------------- | ----------------------------- | -------------- |
+| <code>80c800</code> | <code>{(id=1, dist=0)}</code> | <code>{(id=1, dist=0)}</code> | <code>{(id=1, dist=0)}</code> | #1             |
+| <code>d9c29f</code> | <code>{(id=2, dist=1)}</code> | <code>{(id=1, dist=1)}</code> | <code>{(id=1, dist=1)}</code> | #1             |
+| <code>c85b4b</code> | <code>{(id=2, dist=0)}</code> | <code>{(id=2, dist=0)}</code> | <code>{(id=2, dist=0)}</code> | #2             |
+| <code>69a5ed</code> | <code>{(id=3, dist=2)}</code> | <code>{(id=1, dist=1)}</code> | <code>{(id=1, dist=1)}</code> | #1             |
+| <code>063211</code> | <code>{}</code>               | <code>{(id=1, dist=2)}</code> | <code>{(id=1, dist=2)}</code> | #1             |
+| <code>9d9c37</code> | <code>{}</code>               | <code>{(id=1, dist=3)}</code> | <code>{(id=1, dist=3)}</code> | #1             |
+| <code>f9727d</code> | <code>{(id=3, dist=1)}</code> | <code>{(id=1, dist=2)}</code> | <code>{(id=3, dist=1)}</code> | #3             |
+| <code>3daedb</code> | <code>{(id=3, dist=0)}</code> | <code>{(id=3, dist=0)}</code> | <code>{(id=3, dist=0)}</code> | #3             |
+| <code>e8331f</code> | <code>{}</code>               | <code>{(id=3, dist=1)}</code> | <code>{(id=3, dist=1)}</code> | #3             |
+
+The topological ordering of the commit graph and each traversal takes time linear to the size of the commit graph, making this entire procedure a linear time operation.
+
+## The neglected scaling dimension
+
+Well, it's not _quite_ linear time if you take into account some of the stuff we claimed we could hand-wave away earlier: index files for different root directories.
+
+Many large repositories are built up of smaller, self-contained projects. Or, at least independently analyzable units of code. This enables a fairly corse caching scheme: each time the repository is indexed on git push or periodically, only the units fo code that have had explicitly changed since the last indexed commit needs to be re-indexed.
```

Comment:
```suggestion
Many large repositories are built up of smaller, self-contained projects. Or, at least independently analyzable units of code. This enables a fairly coarse caching scheme: each time the repository is indexed on git push or periodically, only the units fo code that have had explicitly changed since the last indexed commit needs to be re-indexed.
```